Add MOAB-based lateral halo exchange infrastructure in ELM

bishtgautam opened the pull request:
Extends ELM with MOAB-backed MPI halo (ghost cell) exchange to support
lateral transport between columns. All new code is guarded by
`#ifdef HAVE_MOAB`.

- Extend `bounds_type` / `processor_type` with owned, ghost, and
  all-cells index tiers for every subgrid level.
- Set up ghost grid cells and naturally-vegetated ghost columns via
  MOAB ghost-entity determination.
- Add `domainLateralMod` with `oneD_int_data_for_moab` and
  `twoD_real_data_for_moab` tagged-data containers and generic
  `GridLevelIntegerDataHaloExchange` / `GridLevelRealDataHaloExchange`
  APIs.
- Add `ColumnConnectionSetType` to build and store column-to-column
  lateral connection pairs from MOAB internal edges.
- Demonstrate the workflow in `SoilStateType::InitColdGhost`, which
  exchanges soil properties for ghost columns in a single batched MPI
  round.

[BFB]

bishtgautam committed:
Changes the error check because the field may include data for ghost elements

bishtgautam committed:
Initialize only the locally owned cells

bishtgautam committed:
Adds code to do MOAB-based halo exchange

bishtgautam committed:
Adds code to initialize naturally vegetated ghost columns

bishtgautam committed:
Perform halo exhcnage for soil properties

bishtgautam committed:
Initialize ghost grid cells

bishtgautam committed:
Modifies computation of beg/end indices for ghost subgrid units

bishtgautam committed:
Adds column-to-column connections when using MOAB

bishtgautam committed:
elm: Improve MOAB halo exchange in SoilStateType and domainLateralMod

Phase 1 - Add assertion in PackOwnedGridLevelDataForMOAB:
Track column count per grid cell; call endrun if more than
one owned column of the requested type maps to the same grid
cell. This enforces the one-nat-veg-column-per-grid-cell
invariant at the point of packing.

Phase 2 - Decouple halo-exchange API from pre-registered tag:
- Remove soil_lyr_data_real from domainlateral_type and its
  setup call from domainlateral_init.
- Replace GridLevelSoilLayerDataHaloExchange (which relied on
  the pre-registered tag) with GridLevelRealDataHaloExchange
  whose caller supplies and owns the twoD_real_data_for_moab
  struct, mirroring GridLevelIntegerDataHaloExchange.
- Make setup_twoD_real_data_for_moab public so callers can
  initialise their own structs.

Phase 3 - Batch all four fields into one MPI round:
- Add BatchExchangeFieldsUsingMOAB that packs watsat, hksat,
  bsw, and sucsat into a single 4*nlevgrnd-component buffer,
  calls GridLevelRealDataHaloExchange once, then unpacks.
- Replace the four sequential ExchangeAFieldUsingMOAB calls
  in InitColdGhost with one BatchExchangeFieldsUsingMOAB call.
- Delete ExchangeAFieldUsingMOAB (no remaining callers).

Result: MPI communication rounds in InitColdGhost drop from
4 to 1; heap allocations drop from 4 to 1.

peterdschwartz commented on components/elm/src/main/initGridCellsMod.F90:
```diff
@@ -1330,4 +1333,56 @@ end subroutine CheckGhostSubgridHierarchy
 !^ifdef USE_PETSC_LIB
 
+
+#if HAVE_MOAB
```
should be `#ifdef HAVE_MOAB`

bishtgautam replied:
fixed.

vijaysm commented on components/elm/src/utils/domainLateralMod.F90:
```diff
@@ -545,6 +545,295 @@ subroutine ExchangeColumnLevelGhostData(bounds_proc, nvals_per_col, &
+
+    ! get the data from MOAB tag
+    ierr = iMOAB_GetIntTagStorage(data%moab_app_id, data%tag_name, data%ngcells * data%num_comp, data%entity_type(1), data%values)
+    if (ierr > 0) call endrun('Error: setting values in MOAB tag failed.')
```
getting instead of setting

bishtgautam replied:
fixed.

vijaysm commented on components/elm/src/utils/domainLateralMod.F90:
```diff
@@ -545,6 +545,295 @@ subroutine ExchangeColumnLevelGhostData(bounds_proc, nvals_per_col, &
+
+    ! get the data from MOAB tag
+    ierr = iMOAB_GetDoubleTagStorage(data%moab_app_id, data%tag_name, data%ngcells * data%num_comp, data%entity_type(1), data%values)
+    if (ierr > 0) call endrun('Error: setting values in MOAB tag failed.')
```
get instead of set

bishtgautam replied:
fixed.

vijaysm commented on components/elm/src/utils/domainLateralMod.F90:
```diff
@@ -545,6 +545,295 @@ subroutine ExchangeColumnLevelGhostData(bounds_proc, nvals_per_col, &
+    do g = begg, endg_owned
+       idx = moab_gcell%elm2moab(g)
+       do j = 1, data_moab%num_comp
+          data_moab%values(j, idx) = elm_data(g, j)
```
Is g in global index space?

bishtgautam replied:
Yes, `g` is the global space.

vijaysm commented on components/elm/src/main/elm_initializeMod.F90:
```diff
@@ -413,6 +418,10 @@ subroutine initialize1( )
     ! This is needed here for the following call to decompInit_glcp
 
     call initGridCells()
+    call initGhostGridCells()
```
Is lateral coupling optional, or will it always be performed? Just wondering whether this needs to be within `if (lateral_connectivity) then` check

vijaysm commented on components/elm/src/biogeophys/SoilStateType.F90:
```diff
@@ -1080,6 +1084,155 @@ subroutine InitColdGhost(this, bounds_proc)
+    end do
+
+    ! --- single MPI halo exchange ---
+    call setup_twoD_real_data_for_moab(mlndghostid, 'batch_soil_data', nfields*nlevgrnd, &
```
This routine defines the multi-column tag to halo-exchange 4 fields in one shot. The tag that is defined can be reused. Ideally, this should be done once during setup. Except for wasted cycles trying to check whether tags were already defined, this would be technically be a no-op to keep it here.

bishtgautam replied:
@vijaysm, apologies that it took me way too long to get back to this PR. I don't quite understand the issue that your comment is trying to raise.

vijaysm commented on components/elm/src/biogeophys/SoilStateType.F90:
```diff
@@ -1080,6 +1084,155 @@ subroutine InitColdGhost(this, bounds_proc)
+    ! --- single MPI halo exchange ---
+    call setup_twoD_real_data_for_moab(mlndghostid, 'batch_soil_data', nfields*nlevgrnd, &
+                                       moab_gcell%num_ghosted, data_moab)
+    call GridLevelRealDataHaloExchange(data_moab, bounds_proc%begg, bounds_proc%endg, &
```
This is the real halo exchange and should remain here

vijaysm reviewed (changes requested):
Looks good. Let me know if any comments are unclear.

bishtgautam committed:
Address review comments

bishtgautam commented:
@peterdschwartz, I rebased the branch off the latest master, fixed the code, and ran the e3sm_land_developer test suite on pm-cpu. All tests passed. Please re-review and merge this PR, when `next` is open. Thanks.

peterdschwartz commented:
back on next

peterdschwartz commented:
merged to master
